pkg/flow/node.go

```diff
@@ -5,56 +5,117 @@ import (
 	"github.com/jexia/semaphore/pkg/specs"
 	"github.com/jexia/semaphore/pkg/transport"
 	"go.uber.org/zap"
 )
 
+// NodeOptions represent a set of options that could be set through
+// option functions.
+type NodeOptions struct {
+	condition  *Condition
+	call       Call
+	functions  functions.Stack
+	rollback   Call
+	middleware *NodeMiddleware
+}
+
+// NodeOption is a wrapper function
+type NodeOption func(*NodeOptions)
+
+// WithCall sets the given call
+func WithCall(call Call) NodeOption {
+	return func(option *NodeOptions) {
+		option.call = call
+	}
+}
+
+// WithRollback sets the given call as rollback
+func WithRollback(call Call) NodeOption {
+	return func(option *NodeOptions) {
+		option.rollback = call
+	}
+}
+
+// WithCondition sets the given condition
+func WithCondition(condition *Condition) NodeOption {
+	return func(option *NodeOptions) {
+		option.condition = condition
+	}
+}
+
+// WithFunctions sets the given functions stack
+func WithFunctions(functions functions.Stack) NodeOption {
+	return func(option *NodeOptions) {
+		option.functions = functions
+	}
+}
+
+// WithNodeMiddleware sets the given middleware
+func WithNodeMiddleware(middleware *NodeMiddleware) NodeOption {
+	return func(option *NodeOptions) {
+		option.middleware = middleware
+	}
+}
+
+// NewNodeOptions constructs a new node options object and collects the options.
+func NewNodeOptions(opts ...NodeOption) NodeOptions {
+	options := NodeOptions{
+		middleware: &NodeMiddleware{},
+	}
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return options
+}
+
 // NewNode constructs a new node for the given call.
 // The service called inside the call endpoint is retrieved from the services collection.
 // The call, codec and rollback are defined inside the node and used while processing requests.
-func NewNode(parent *broker.Context, node *specs.Node, condition *Condition, call, rollback Call, middleware *NodeMiddleware) *Node {
+func NewNode(parent *broker.Context, node *specs.Node, opts ...NodeOption) *Node {
+	options := NewNodeOptions(opts...)
+
 	module := broker.WithModule(parent, "node", node.Name)
 	ctx := logger.WithLogger(module)
 
-	refs := references.References{}
-
-	if middleware == nil {
-		middleware = &NodeMiddleware{}
-	}
+	refs := references.Collection{}
 
 	if node.Call != nil {
 		refs.MergeLeft(references.ParameterReferences(node.Call.Request))
 	}
 
-	if call != nil {
-		for _, prop := range call.References() {
+	if options.call != nil {
+		for _, prop := range options.call.References() {
 			refs.MergeLeft(references.PropertyReferences(prop))
 		}
 	}
 
-	if rollback != nil {
-		for _, prop := range rollback.References() {
+	if options.rollback != nil {
+		for _, prop := range options.rollback.References() {
 			refs.MergeLeft(references.PropertyReferences(prop))
 		}
 	}
 
 	return &Node{
-		BeforeDo:     middleware.BeforeDo,
-		BeforeRevert: middleware.BeforeRollback,
-		Condition:    condition,
+		BeforeDo:     options.middleware.BeforeDo,
```

Comment:
does it make sense to store the entire `Options` object inside `Node` instead of having `BeforeDo`, `BeforeRevert`, ... ?

Reply:
Yes, I wanted to include this inside the next refactor but could include this inside this PR as well.